From d3c5a180bcd1cfdb72f134606e820ac8dde38037 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Wed, 30 Nov 2011 12:32:30 -0700
Subject: [PATCH] Add decoration regions to the Screen header title

This is primarily a feature to support moving the
review button into the title line on the diff screen.
This change does not actually move the button.

The screen header is mainly a text title.  Previously
a widget could only be inserted to the left of the
title, this was done for the star on the change
screen.

Enhance the title "decoration" functionality so it
can be decorated in the West, the East, and the FarEast
by any Widget. The West and East decorations will
'surround" the text on the left and right respectively,
and the FarEast will be right justified to the right
edge of the screen.  The East decoration will expand to
take up any extra space.

Change-Id: Iaa6b7775ae16bd94e24c5d54ad9e8c029a7893bb
---
 .../google/gerrit/client/changes/ChangeScreen.java |    2 +-
 .../java/com/google/gerrit/client/ui/Screen.java   |   48 +++++++++++++++++--
 2 files changed, 44 insertions(+), 6 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
index ed0f2f2..1b382a8 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
@@ -204,7 +204,7 @@ public class ChangeScreen extends Screen {
           toggleStar();
         }
       });
-      insertTitleWidget(starChange);
+      setTitleWest(starChange);
     }
 
     descriptionBlock = new ChangeDescriptionBlock();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/Screen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/Screen.java
index 68026ae..b22cae5 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/Screen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/Screen.java
@@ -18,12 +18,24 @@ import com.google.gerrit.client.Gerrit;
 import com.google.gerrit.common.PageLinks;
 import com.google.gwt.user.client.History;
 import com.google.gwt.user.client.ui.FlowPanel;
+import com.google.gwt.user.client.ui.Grid;
+import com.google.gwt.user.client.ui.HasHorizontalAlignment;
 import com.google.gwt.user.client.ui.InlineLabel;
 import com.google.gwt.user.client.ui.Widget;
 import com.google.gwtexpui.user.client.View;
 
+/**
+  *  A Screen layout with a header and a body.
+  *
+  * The header is mainly a text title, but it can be decorated
+  * in the West, the East, and the FarEast by any Widget.  The
+  * West and East decorations will 'surround" the text on the
+  * left and right respectively, and the FarEast will be right
+  * justified to the right edge of the screen.  The East
+  * decoration will expand to take up any extra space.
+  */
 public abstract class Screen extends View {
-  private FlowPanel header;
+  private Grid header;
   private InlineLabel headerText;
   private FlowPanel body;
   private String token;
@@ -46,13 +58,31 @@ public abstract class Screen extends View {
   public void registerKeys() {
   }
 
+  private static enum Cols {
+    West(0), Title(1), East(2), FarEast(3);
+
+    public final int num;
+
+    Cols(int col) {
+      num = col;
+    }
+  }
+
   protected void onInitUI() {
     final FlowPanel me = (FlowPanel) getWidget();
-    me.add(header = new FlowPanel());
+    me.add(header = new Grid(1, Cols.values().length));
     me.add(body = new FlowPanel());
 
+    FlowPanel title = new FlowPanel();
+    title.add(headerText = new InlineLabel());
+    title.setStyleName(Gerrit.RESOURCES.css().screenHeader());
+    header.setWidget(0, Cols.Title.num, title);
+
     header.setStyleName(Gerrit.RESOURCES.css().screenHeader());
-    header.add(headerText = new InlineLabel());
+    header.getCellFormatter().setHorizontalAlignment(0, Cols.FarEast.num,
+      HasHorizontalAlignment.ALIGN_RIGHT);
+    // force FarEast all the way to the right
+    header.getCellFormatter().setWidth(0, Cols.FarEast.num, "99%");
   }
 
   protected void setWindowTitle(final String text) {
@@ -73,8 +103,16 @@ public abstract class Screen extends View {
     }
   }
 
-  protected void insertTitleWidget(final Widget w) {
-    header.insert(w, 0);
+  protected void setTitleEast(final Widget w) {
+    header.setWidget(0, Cols.East.num, w);
+  }
+
+  protected void setTitleFarEast(final Widget w) {
+    header.setWidget(0, Cols.FarEast.num, w);
+  }
+
+  protected void setTitleWest(final Widget w) {
+    header.setWidget(0, Cols.West.num, w);
   }
 
   protected void add(final Widget w) {
-- 
1.7.6

